src/main/java/org/apache/commons/compress/utils/IOUtils.java

```diff
@@ -275,6 +277,49 @@ public static byte[] readRange(final ReadableByteChannel input, final int len) t
         return output.toByteArray();
     }
 
+    /**
+     * Writes full buffer to channel.
+     *
+     * @param channel
+     *      channel to write to
+     * @param buf
+     *      buffer to write
+     * @throws IOException
+     *      when writing fails
+     */
+    public static void writeFully(SeekableByteChannel channel, ByteBuffer buf) throws IOException {
+        while (buf.hasRemaining()) {
+            int remaining = buf.remaining();
+            int written = channel.write(buf);
+            if (written <= 0) {
+                throw new IOException("Failed to fully write: channel=" + channel + " length=" + remaining + " written=" + written);
+            }
+        }
+    }
+
+    /**
+     * Writes full buffer to channel at specified position.
+     *
+     * @param channel
+     *      channel to write to
+     * @param buf
+     *      buffer to write
+     * @param position
+     *      position to write at
+     * @throws IOException
+     *      when writing fails
+     */
+    public static void writeFullyAt(FileChannel channel, ByteBuffer buf, long position) throws IOException {
```

Comment:
See above. All new public and protected elements shoyuld have an `@since` Javadoc tag.

Reply:
Moved to package private ZipIoUtil as requested above.